packages/next/src/server/app-render/required-scripts.tsx

```diff
@@ -0,0 +1,56 @@
+import type { BuildManifest } from '../get-page-files'
+
+import ReactDOM from 'react-dom'
+
+export function getRequiredScripts(
+  buildManifest: BuildManifest,
+  assetPrefix: string,
+  SRIManifest: undefined | Record<string, string>,
+  qs: string
+): [() => void, string | { src: string; integrity: string }] {
+  let preinitScripts: () => void
+  let preinitScriptCommands: string[] = []
+  let bootstrapScript: string | { src: string; integrity: string } = ''
+  const files = buildManifest.rootMainFiles
+  if (files.length === 0) {
+    throw new Error(
+      'Invariant: missing bootstrap script. This is a bug in Next.js'
+    )
+  }
+  if (SRIManifest) {
+    bootstrapScript = {
+      src: `${assetPrefix}/_next/` + files[0] + qs,
+      integrity: SRIManifest[files[0]],
+    }
+    for (let i = 1; i < files.length; i++) {
+      const src = `${assetPrefix}/_next/` + files[i] + qs
```

Comment:
Nit pick: since this is used four times, you could keep it [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) with a shared function:

```ts
const getSrc = (i: number) => `${assetPrefix}/_next/` + files[i] + qs
```

Then you can `getSrc(0)` or `getSrc(i)` later

Reply:
We don't bundle the server code all the time yet (I think) and I'm not sure what level of optimization we will end up applying to our bundle when we do. Absent an optimizing step that would inline these function calls it's more performant to just inline the string concatenation.

I realize this is a drop in the bucket compared to overhead in the rest of Next but I think in the long run we're going to want to start caring more about good perf by default and in aggregate little things like this will add up eventually.

There are a couple reasons this would be slower, first a function call itself has overhead. But second the allocation of a new function to create the closure for `assetPrefix` and `qs`. There's obviously a tradeoff (getRequireScripts itself is a function that won't be inlined) but since this repeated string concat is clear in it's intention and self contained I think the repetition is not particularly taxing mental cost and it's not a place we're likely to mistakenly update in a broken way in the future.

I could make a `getSrc(prefix, filename, querystring)` function that is only allocated once for the module but then it's not really saving much character-wise and still adds a function call overhead so I think it's good how it is